arch-design-questions.md

```diff
@@ -0,0 +1,93 @@
+# Client design
+
+## Principles
+
+- There are two top level clients - `ServiceBusReceiverClient` and `ServiceBusSenderClient`.
+- Clients are [single-use](#single-use-clients) - they are created with a receive mode, session ID (if applicable), queue/topic-subscription
+  baked in and cannot be changed.
+- The user can choose between two different methods of message delivery - push (via registered message handlers)
+  or pull (via an iterator/reactor pattern or .receiveBatch()-style)
+- Sessions have two different methods for interacting with them:
+  1. [single-use](#single-use-clients)
+  2. [session manager](#sessions-design)
+- Different message types will be used for sending messages vs receiving messages. This is less confusing for
+  the user since we can remove fields that are non-sensical for sending (for instance, `locktoken`).
+
+# Single use clients
+
+Clients are single-use - you cannot change their receive mode or targeted queue/topic/subscription after
+they are created.
+
+This also means that calling .receive() (or registering multiple handlers) isn't a useful activity - the user
+would only be competing against _themselves_ since calling receive() multiple times would just result
+in multiple links, in the same mode, to the same entity, on the same _connection_.
+
+# Sessions design
+
+The primary use case for non-session bus users when receiving is opening
+up a single client instance, connecting to a targeted queue (or topic/subscription)
+and receiving from it.
+
+In this case the lifetime of the client is essentially the lifetime of their application which makes
+connection management a non-issue.
+
+Sessions differ in that the user's application is not tied to a single session - the user
+is expected to open up and read from many sessions during the lifetime of their application.
+
+This requires a few design considerations:
+
+## Connection "explosion"
+
+**Problem**
+
+The user opening multiple instances of a client means we need a way to limit the number of
+physical connections to service bus.
+
+Opening multiple receive _links_ is fine with service bus (it's a cheap abstraction within
```

Comment:
Is this really an important distinction at the architectural level to discuss sharing connections between clients?   Seems like something that would be addressed verbally if a question comes up in discussions.

Reply:
This doc is the other direction - we're discussing things that came up _during_ the architecture meeting that we, as a feature crew, need to discuss and answer. It's definitely a detail we should care about.